From de469e025cbf75191459127f409db08e4ff95d2e Mon Sep 17 00:00:00 2001
From: Arnon Kaufman <akaufman@habana.ai>
Date: Sun, 22 Oct 2023 18:15:40 +0300
Subject: [PATCH] [SW-15774] build: refactor zephyr package and scripts

For backward compatible, Zephyr's package's build script needs to be
refactor, and no longer leads the build process, but rather use as a
consolidator of individual packages. It is done by adding the relevant
packages as dependencies.

The second goal of this patch is to update the location and names of
intermediate files.

Change-Id: I85f381745784603b1ce2bbe561f4a9dc6ded73e5
Signed-off-by: Arnon Kaufman <akaufman@habana.ai>
---

diff --git a/board/habanalabs/common/post-image-fit-global.sh b/board/habanalabs/common/post-image-fit-global.sh
index 7fa938d..d07a7ac 100755
--- a/board/habanalabs/common/post-image-fit-global.sh
+++ b/board/habanalabs/common/post-image-fit-global.sh
@@ -34,8 +34,8 @@
 HL_SEC_KEY_INDEX=`grep HL_SEC_KEY_INDEX $HL_SEC_FILE | sed 's/^.*\(0x.*\)$/\1/'`
 HL_SEC_FW_VERSION=`grep HL_SEC_FW_VERSION $HL_SEC_FILE | awk '{ print $NF }'`
 # set zephyr apps version
-HL_ZEPHYR_MGMT_VERSION=`cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-mgmt/buildversion-habanalabs`
-HL_ZEPHYR_PID_VERSION=`cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-pid/buildversion-habanalabs`
+HL_ZEPHYR_MGMT_VERSION=`cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-mgmt-custom/buildversion-habanalabs`
+HL_ZEPHYR_PID_VERSION=`cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-pid-custom/buildversion-habanalabs`
 
 COMMON_BASE=${BR2_EXTERNAL}/board/habanalabs/common
 UBOOT_ZEPHYR_ITS_FILE=uboot-zephyr-${DEV_NAME}_${HL_TARGET}.its
diff --git a/board/habanalabs/common/post-image-spi-arc_global.sh b/board/habanalabs/common/post-image-spi-arc_global.sh
index 2299ab2..0117fe6 100755
--- a/board/habanalabs/common/post-image-spi-arc_global.sh
+++ b/board/habanalabs/common/post-image-spi-arc_global.sh
@@ -21,9 +21,9 @@
 PIF_ICCM_BINFILE=${PIF_ICCM_BINFILE:-pcie_pif-iccm.bin}
 PIF_DCCM_BINFILE=${PIF_DCCM_BINFILE:-pcie_pif-dccm.bin}
 
-ZEPHYR_PID_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-pid
-ZEPHYR_MGMT_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-mgmt
-ZEPHYR_EVNT_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-event
+ZEPHYR_PID_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-pid-custom
+ZEPHYR_MGMT_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-mgmt-custom
+ZEPHYR_EVNT_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-event-custom
 
 PRIME_IMAGE=0
 SECOND_IMAGE=1
diff --git a/board/habanalabs/gaudi2/post-image-agents.sh b/board/habanalabs/gaudi2/post-image-agents.sh
index a3eeac5..e423f0f 100755
--- a/board/habanalabs/gaudi2/post-image-agents.sh
+++ b/board/habanalabs/gaudi2/post-image-agents.sh
@@ -25,7 +25,7 @@
 AGENT_ARCH="arm64"
 LOADABLES="security\", \"ppboot\", \"preboot\", \"img-ptr"
 
-HL_PREBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}_arc/build/zephyr-custom/build-preboot/buildversion-habanalabs`
+HL_PREBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}_arc/build/zephyr-preboot-custom/build/buildversion-habanalabs`
 HL_PPBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}_arc/build/preboot-custom/buildversion-habanalabs | cut -d ' ' -f1`
 HL_FW_VERSION=`cut -d- -f2- ${BASE_DIR}/build/linux-custom/localversion-habanalabs`
 # Set hl-*-version = "hl-<device>-<HL_VERSION>[-rc]-fw-<fw ver>[-<git data>]"; inside FIT header
diff --git a/board/habanalabs/gaudi3/post-image-fit.sh b/board/habanalabs/gaudi3/post-image-fit.sh
index 974dcb6..b7150d0 100644
--- a/board/habanalabs/gaudi3/post-image-fit.sh
+++ b/board/habanalabs/gaudi3/post-image-fit.sh
@@ -59,9 +59,9 @@
 	hl_extract_zephyr_elf_sections
 popd
 
-MGMT_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/build-mgmt/zephyr/zephyr.stat
-PID_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/build-pid/zephyr/zephyr.stat
-EVENT_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/build-event/zephyr/zephyr.stat
+MGMT_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-mgmt-custom/build/zephyr/zephyr.stat
+PID_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-pid-custom/build/zephyr/zephyr.stat
+EVENT_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-event-custom/build/zephyr/zephyr.stat
 
 # Extract the start address from stat file
 ENTRY_ADDRESS=$(cat $MGMT_STAT | grep "Entry point address" | head -1)
diff --git a/board/habanalabs/gaudi3_arc/post-image-agents.sh b/board/habanalabs/gaudi3_arc/post-image-agents.sh
index 11b6705..1a281bc 100644
--- a/board/habanalabs/gaudi3_arc/post-image-agents.sh
+++ b/board/habanalabs/gaudi3_arc/post-image-agents.sh
@@ -8,8 +8,10 @@
 source ${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/create-fake-fuse-image.sh
 
 HL_COMPRESS=none
-HL_PREBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/build-preboot/buildversion-habanalabs`
-HL_PPBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}/build/preboot-custom/buildversion-habanalabs | cut -d ' ' -f1`
+#HL_PREBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}_arc/build/zephyr-preboot-custom/buildversion-habanalabs`
+#HL_PPBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}_arc/build/zephyr-preboot-custom/buildversion-habanalabs | cut -d ' ' -f1`
+HL_PREBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-preboot-custom/build/buildversion-habanalabs`
+HL_PPBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-preboot-custom/build/buildversion-habanalabs | cut -d ' ' -f1`
 HL_UBOOT_VERSION=""
 HL_FW_VERSION=""
 GAUDI3_SRAM_ADDR_LO=0xF0000000
@@ -20,7 +22,7 @@
 AGENT_ARCH="arc"
 LOADABLES="ppboot\", \"preboot\", \"img-ptr\", \"fake-fuse-image\", \"pif-iccm\", \"pif-dccm"
 
-BIN=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/build-agent-fw-loader/zephyr/zephyr*.bin
+BIN=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-agent-fwl-custom/build/zephyr/zephyr*.bin
 GAUDI3_AGENT_BIN=$(ls ${BIN} 2>/dev/null)
 GAUDI3_AGENT_STAT=$(echo $GAUDI3_AGENT_BIN | sed 's/\.bin/.stat/')
 GAUDI3_SPEC_FW_LOADER_ITS=$(ls ${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/agent-fw_loader.its)
diff --git a/package/zephyr/gaudi2.mk b/package/zephyr/gaudi2.mk
new file mode 100644
index 0000000..bcdb2e6
--- /dev/null
+++ b/package/zephyr/gaudi2.mk
@@ -0,0 +1,11 @@
+#Buildroot is including all the makefiles for
+#the compilation of zephyr appliation Hence
+#adding additional check for gaudi3 to avoid
+#including SPECS_H9_DIR for other platforms
+
+GAUDI2_CONF_DEV_EXTRA_CFLAGS =\
+	-I$(SPECS_EMBEDDED_DIR) \
+	-I$(SPECS_EMBEDDED_DIR)/$(DEV_PREFIX) \
+	-I$(SPECS_ROOT) \
+	-I$(SPECS_EXT_ROOT) \
+	-I$(SPECS_ROOT)/$(DEV_PREFIX)/asic_reg
diff --git a/package/zephyr/gaudi3.mk b/package/zephyr/gaudi3.mk
index 194f786..98774be 100644
--- a/package/zephyr/gaudi3.mk
+++ b/package/zephyr/gaudi3.mk
@@ -3,7 +3,11 @@
 #adding additional check for gaudi3 to avoid
 #including SPECS_H9_DIR for other platforms
 
-ifeq ($(DEV_PREFIX),gaudi3)
-ZEPHYR_CONF_DEV_EXTRA_CFLAGS =  -I$(SPECS_H9_DIR)
-ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -I$(SPECS_H9_DIR)/gaudi3/asic_reg
-endif
+GAUDI3_CONF_DEV_EXTRA_CFLAGS =\
+	-I$(SPECS_EMBEDDED_DIR) \
+	-I$(SPECS_EMBEDDED_DIR)/$(DEV_PREFIX) \
+	-I$(SPECS_ROOT) \
+	-I$(SPECS_EXT_ROOT) \
+	-I$(SPECS_ROOT)/$(DEV_PREFIX)/asic_reg \
+	-I$(SPECS_H9_DIR) \
+	-I$(SPECS_H9_DIR)/$(DEV_PREFIX)/asic_reg
diff --git a/package/zephyr/zephyr.mk b/package/zephyr/zephyr.mk
index 300525d..c6d4faa 100644
--- a/package/zephyr/zephyr.mk
+++ b/package/zephyr/zephyr.mk
@@ -4,264 +4,48 @@
 
 ZEPHYR_DEPENDENCIES = host-cmake host-lzop host-lz4
 
-ZEPHYR_CONF_ENV = CROSS_COMPILE=$(TARGET_CROSS)
-ZEPHYR_CONF_ENV += ZEPHYR_TOOLCHAIN_VARIANT=cross-compile
-ZEPHYR_CONF_ENV += ZEPHYR_BASE=$(ZEPHYR_BUILDDIR)
-
-ZEPHYR_CONF_OPTS = -UCMAKE_TOOLCHAIN_FILE
-ZEPHYR_CONF_OPTS += -DUSER_CACHE_DIR=$(@D)/.cache
-ZEPHYR_CONF_OPTS += -DCMAKE_EXE_LINKER_FLAGS_INIT=-Wl,-e__start
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-Werror -Wno-misleading-indentation $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-ZEPHYR_CONF_OPTS += -DBUILD_VERSION="$(shell cut -d- -f2- $(ZEPHYR_BUILDDIR)/localversion-habanalabs)"
-ZEPHYR_CONF_OPTS += -DBUILD_MACHINE="$(shell hostname -s)"
-ZEPHYR_CONF_OPTS += -DCMAKE_C_FLAGS="-I$(SPECS_EMBEDDED_DIR) -I$(SPECS_EMBEDDED_DIR)/$(DEV_PREFIX) -I$(SPECS_ROOT) -I$(SPECS_EXT_ROOT) -I$(SPECS_ROOT)/$(DEV_PREFIX)/asic_reg"
-ZEPHYR_CONF_OPTS += -DDTS_EXTRA_CPPFLAGS="-I$(SPECS_EMBEDDED_DIR)"
-
-DEVICE_NAME = $(firstword $(subst _, ,$(HL_DEV)))
-
-# zephyr uses west tool to perform repo management, installation
-# of the tools used for zephyr compilation. For our zephyr
-# compilation, we do not make use of west tool.
-# If west is installed on build machine, it looks for configuration
-# file and throws error if it doesn't find one. To solve this issue,
-# create dummy configuration file in .west folder relative to zephyr
-# build directory.
-ZEPHYR_WEST_SRC_FILE_PATH = $(BR2_EXTERNAL_HABANALABS_PATH)/package/zephyr
-ZEPHYR_WEST_DEST_DIR_PATH = $(BR2_EXTERNAL_HABANALABS_PATH)/output/$(HL_DEV)/build/.west
-
-ZEPHYR_LOCAL_VERSION = $$(sed "s/^.*fw-\\([0-9]\\+\\.[0-9]\\+\\.[0-9]\\+\\).*$$/\\1/" $(ZEPHYR_BUILDDIR)/localversion-habanalabs)
-
-ZEPHYR_CONFIGURE_CMDS = $(INSTALL) -D -m 0755 $(ZEPHYR_WEST_SRC_FILE_PATH)/west.config $(ZEPHYR_WEST_DEST_DIR_PATH)/config ; cd $(ZEPHYR_BUILDDIR) ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#ifndef _HL_VERSION_" > include/hl_version.h ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#define _HL_VERSION_" >> include/hl_version.h ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#define HL_VERSION_MAJOR $$(echo $(ZEPHYR_LOCAL_VERSION) | cut -d. -f1)" >> $(ZEPHYR_BUILDDIR)/include/hl_version.h ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#define HL_VERSION_MINOR $$(echo $(ZEPHYR_LOCAL_VERSION) | cut -d. -f2)" >> $(ZEPHYR_BUILDDIR)/include/hl_version.h ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#define HL_VERSION_PATCH $$(echo $(ZEPHYR_LOCAL_VERSION) | cut -d. -f3)" >> $(ZEPHYR_BUILDDIR)/include/hl_version.h ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#endif" >> include/hl_version.h
-ZEPHYR_BUILD_CMDS = true
-ZEPHYR_INSTALL_TARGET_CMDS = true
-
-ZEPHYR_COMPONENTS_LIST =
-
-# Arguments:
-# - module name
-# - component dir
-# - component name
-define DEFINE_MODULE_ENVVARS
-	$(eval MODULE_NAME=$(strip $1)) \
-	$(eval MODULE_NAME_UPPER_CASE=$(shell echo $(MODULE_NAME) | tr a-z A-Z)) \
-	$(eval MODULE_NAME_UPPER_CASE=$(strip $(MODULE_NAME_UPPER_CASE))) \
-	$(eval COMPONENT_NAME_UPPER_CASE=$(shell echo $3 | tr a-z A-Z)) \
-	$(eval COMPONENT_NAME_UPPER_CASE=$(strip $(COMPONENT_NAME_UPPER_CASE))) \
-	\
-	$(eval $(COMPONENT_NAME_UPPER_CASE)_$(MODULE_NAME_UPPER_CASE)_SRCDIR=$(EMBEDDED_DIR)/zephyr-modules/$(MODULE_NAME)) \
-	$(eval $(COMPONENT_NAME_UPPER_CASE)_$(MODULE_NAME_UPPER_CASE)_DIR=$2/modules/$(MODULE_NAME)) \
-	$(eval $(COMPONENT_NAME_UPPER_CASE)_$(MODULE_NAME_UPPER_CASE)_NAME=$3_$(MODULE_NAME))
-endef
-
-define CREATE_APP_VERSION
--DAPP_NAME=$1 \
--DAPP_VERSION=$(shell cut -d- -f2- $(ZEPHYR_BUILDDIR)/$2/localversion-habanalabs)
-endef
-
 ifdef BR2_PACKAGE_ZEPHYR_ARC_PID
-ZEPHYR_COMPONENTS_LIST += pid
-ZEPHYR_PID_DIR = build-pid
-ZEPHYR_PID_NAME = zephyr_arc_pid
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_PID_DIR)
-ZEPHYR_CONF_OPTS_PID += -DBOARD=hl_$(DEVICE_NAME)_pid_$(HL_TARGET)
-ZEPHYR_CONF_OPTS_PID += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_PID_SRCDIR = $(EMBEDDED_DIR)/zephyr-apps/arcpid
-ZEPHYR_CONF_OPTS_PID += $(ZEPHYR_PID_SRCDIR)
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,arcpid,$(ZEPHYR_PID_DIR)) $(ZEPHYR_CONF_OPTS_PID) -B $(ZEPHYR_PID_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_PID_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_PID_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_pid.elf
-ZEPHYR_PID_MODULES = mbedtls boot_info eeprom
-$(foreach module,$(ZEPHYR_PID_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_PID_DIR),$(ZEPHYR_PID_NAME)))
+ZEPHYR_DEPENDENCIES += zephyr-pid
 endif
 
 ifdef BR2_PACKAGE_ZEPHYR_ARC_MGMT
-ZEPHYR_COMPONENTS_LIST += mgmt
-ZEPHYR_MGMT_DIR = build-mgmt
-ZEPHYR_MGMT_NAME = zephyr_arc_mgmt
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_MGMT_DIR)
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-Werror -Wno-misleading-indentation -include $(EMBEDDED_ROOT)/specs/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-ZEPHYR_CONF_OPTS_MGMT += -DBOARD=hl_$(DEVICE_NAME)_mgmt_$(HL_TARGET)
-ZEPHYR_CONF_OPTS_MGMT += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_MGMT_SRCDIR = $(EMBEDDED_DIR)/zephyr-apps/arcmgmt
-ZEPHYR_CONF_OPTS_MGMT += $(ZEPHYR_MGMT_SRCDIR)
-ifeq ($(HL_TARGET),pldm)
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-Werror -Wno-misleading-indentation -DTARGET_IS_PLDM -include $(EMBEDDED_ROOT)/specs/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-ifeq ($(PLDM_SKIP_UBOOT),1)
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-Werror -Wno-misleading-indentation -DTARGET_IS_PLDM_REGRESSION -include $(EMBEDDED_ROOT)/specs/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-endif
-else ifeq ($(HL_TARGET),onix)
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-Werror -Wno-misleading-indentation -DTARGET_IS_ONIX -include $(EMBEDDED_ROOT)/specs/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-endif
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,arcmgmt,$(ZEPHYR_MGMT_DIR)) $(ZEPHYR_CONF_OPTS_MGMT) -B $(ZEPHYR_MGMT_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_MGMT_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_MGMT_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_mgmt.elf
-ZEPHYR_MGMT_MODULES = eeprom mbedtls boot_info i2c-indirect soc_init hl_shell
-$(foreach module,$(ZEPHYR_MGMT_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_MGMT_DIR),$(ZEPHYR_MGMT_NAME)))
+ZEPHYR_DEPENDENCIES += zephyr-mgmt
 endif
 
 ifdef BR2_PACKAGE_ZEPHYR_ARC_EVENT
-ZEPHYR_COMPONENTS_LIST += event
-ZEPHYR_EVENT_DIR = build-event
-ZEPHYR_EVENT_NAME = zephyr_arc_event
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_EVENT_DIR)
-ZEPHYR_CONF_OPTS_EVENT += -DBOARD=hl_$(DEVICE_NAME)_event_$(HL_TARGET)
-ZEPHYR_CONF_OPTS_EVENT += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_EVENT_SRCDIR = $(EMBEDDED_DIR)/zephyr-apps/arcevent
-ZEPHYR_CONF_OPTS_EVENT += $(ZEPHYR_EVENT_SRCDIR)
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(ZEPHYR_CONF_OPTS_EVENT) -B $(ZEPHYR_EVENT_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_EVENT_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_EVENT_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_event.elf
-ZEPHYR_EVENT_MODULES = boot_info mbedtls
-$(foreach module,$(ZEPHYR_EVENT_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_EVENT_DIR),$(ZEPHYR_EVENT_NAME)))
+ZEPHYR_DEPENDENCIES += zephyr-event
 endif
 
 ifdef BR2_PACKAGE_ZEPHYR_ARC_PREBOOT
-ZEPHYR_COMPONENTS_LIST += preboot
-ZEPHYR_PREBOOT_DIR = build-preboot
-ZEPHYR_PREBOOT_NAME = zephyr_preboot
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_PREBOOT_DIR)
-ZEPHYR_CONF_OPTS_PREBOOT += -DBOARD=hl_$(DEVICE_NAME)_preboot_$(HL_TARGET)
-ZEPHYR_CONF_OPTS_PREBOOT += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_CONF_OPTS_PREBOOT += -DZEPHYR_BUILD_DIR=$(ZEPHYR_BUILDDIR)
-ZEPHYR_CONF_OPTS_PREBOOT += -DTARGET=$(HL_TARGET)
-ZEPHYR_PREBOOT_SRCDIR = $(EMBEDDED_DIR)/zephyr-apps/preboot
-ZEPHYR_CONF_OPTS_PREBOOT += $(ZEPHYR_PREBOOT_SRCDIR)
-ifeq ($(PREBOOT_RUN_FROM_ICCM),true)
-ZEPHYR_CONF_OPTS_PREBOOT += -DCONFIG_PREBOOT_RUN_FROM_ICCM=y
-endif #PREBOOT_RUN_FROM_ICCM
-#TODO - TARGET_IS_ASIC: Need to check why on ZEPHYR_CONF_OPTS its not working as expected
-ifeq ($(HL_TARGET),asic)
-ZEPHYR_CONF_OPTS_PREBOOT += -DEXTRA_CFLAGS="-DTARGET_IS_ASIC -include $(EMBEDDED_ROOT)/specs/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-endif
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,preboot,$(ZEPHYR_PREBOOT_DIR)) $(ZEPHYR_CONF_OPTS_PREBOOT) -B $(ZEPHYR_PREBOOT_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_PREBOOT_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_PREBOOT_DIR)/zephyr/zephyr.bin $(BINARIES_DIR)/preboot.bin
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_PREBOOT_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_preboot.elf
-ZEPHYR_PREBOOT_MODULES = fdt boot_info mbedtls eeprom soc_init hl_shell
-$(foreach module,$(ZEPHYR_PREBOOT_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_PREBOOT_DIR),$(ZEPHYR_PREBOOT_NAME)))
+ZEPHYR_DEPENDENCIES += zephyr-preboot
 endif #BR2_PACKAGE_ZEPHYR_ARC_PREBOOT
 
-
 ifdef BR2_PACKAGE_ZEPHYR_AGENT_FW_LOADER
-ZEPHYR_COMPONENTS_LIST += fw_loader
-ZEPHYR_FW_LOADER_DIR = build-agent-fw-loader
-ZEPHYR_FW_LOADER_NAME = zephyr_agent_fw_loader
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_FW_LOADER_DIR)
-ifdef BR2_arcle
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-Werror -Wno-misleading-indentation -include $(EMBEDDED_ROOT)/specs/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-ZEPHYR_CONF_OPTS_AGENT_FW_LOADER += -DBOARD=hl_$(DEVICE_NAME)_agent_$(HL_TARGET)
+ZEPHYR_DEPENDENCIES += zephyr-agent-fwl
 endif
-ifdef BR2_aarch64
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-mstrict-align -Werror -Wno-misleading-indentation -include $(EMBEDDED_ROOT)/specs/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-ZEPHYR_CONF_OPTS_AGENT_FW_LOADER += -DBOARD=hl_$(DEVICE_NAME)_$(HL_TARGET)
-endif
-ZEPHYR_CONF_OPTS_AGENT_FW_LOADER += -DDEVICE=$(DEVICE_NAME)
-ifeq ($(HL_TARGET),pldm)
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-Werror -Wno-misleading-indentation -DTARGET_IS_PLDM -include $(EMBEDDED_ROOT)/specs/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-else ifeq ($(HL_TARGET),onix)
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-Werror -Wno-misleading-indentation -DTARGET_IS_ONIX -include $(EMBEDDED_ROOT)/specs/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-endif
-ZEPHYR_FW_LOADER_SRCDIR += $(EMBEDDED_DIR)/zephyr-plugins/fw-loader
-ZEPHYR_CONF_OPTS_AGENT_FW_LOADER += $(ZEPHYR_FW_LOADER_SRCDIR)
-ZEPHYR_CONF_OPTS_AGENT_FW_LOADER += -DFAKE_SECURITY=$(FAKE_SECURITY)
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,fw-loader,$(ZEPHYR_FW_LOADER_DIR)) $(ZEPHYR_CONF_OPTS_AGENT_FW_LOADER) -B $(ZEPHYR_FW_LOADER_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_FW_LOADER_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_FW_LOADER_DIR)/zephyr/zephyr.bin $(BINARIES_DIR)/zephyr_agent_fw_loader.bin
-# The stat file is used for extracting the binary's load address and start address
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_FW_LOADER_DIR)/zephyr/zephyr.stat $(BINARIES_DIR)/zephyr_agent_fw_loader.stat
-ZEPHYR_FW_LOADER_MODULES = boot_info eeprom fdt i2c-indirect lz4 mbedtls hl_shell soc_init
-$(foreach module,$(ZEPHYR_FW_LOADER_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_FW_LOADER_DIR),$(ZEPHYR_FW_LOADER_NAME)))
-endif
-
-# first argument is agent name, second argument is additional configuration opts
-define CREATE_ZEPHYR_MARGIN_TOOL_AGENT
-$(eval AGENT_LOWER_CASE = $(strip $1)) \
-$(eval ZEPHYR_COMPONENTS_LIST += $(AGENT_LOWER_CASE)) \
-$(eval AGENT_UPPER_CASE = $(shell echo $(AGENT_LOWER_CASE) | tr a-z A-Z)) \
-$(eval AGENT_UPPER_CASE = $(strip $(AGENT_UPPER_CASE))) \
-$(eval ZEPHYR_AGENT_DIR = build-agent-$1) \
-$(eval ZEPHYR_$(AGENT_UPPER_CASE)_DIR = build-agent-$(AGENT_LOWER_CASE)) \
-$(eval ZEPHYR_$(AGENT_UPPER_CASE)_NAME = zephyr_agent_$(AGENT_LOWER_CASE)) \
-$(eval ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_AGENT_DIR)) \
-$(eval ZEPHYR_CONF_OPTS_AGENT = -DBOARD=hl_$(DEVICE_NAME)_$(HL_TARGET)) \
-$(eval ZEPHYR_CONF_OPTS_AGENT += -DDEVICE=$(DEVICE_NAME)) \
-$(eval ZEPHYR_$(AGENT_UPPER_CASE)_SRCDIR = $(EMBEDDED_DIR)/zephyr-plugins/margin-tools/$(AGENT_LOWER_CASE)) \
-$(eval ZEPHYR_CONF_OPTS_AGENT += $(ZEPHYR_$(AGENT_UPPER_CASE)_SRCDIR)) \
-$(eval ZEPHYR_CONF_OPTS_AGENT += $2) \
-$(eval ZEPHYR_CONFIGURE_CMDS += && $$(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $$(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,$(AGENT_LOWER_CASE),$(ZEPHYR_AGENT_DIR)) $(ZEPHYR_CONF_OPTS_AGENT) -B $(ZEPHYR_AGENT_DIR)) \
-$(eval ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $$(@D)/$(ZEPHYR_AGENT_DIR)) \
-$(eval ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $$(@D)/$(ZEPHYR_AGENT_DIR)/zephyr/zephyr.bin $(BINARIES_DIR)/zephyr_agent_$1.bin) \
-$(eval ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $$(@D)/$(ZEPHYR_AGENT_DIR)/zephyr/zephyr.stat $(BINARIES_DIR)/zephyr_agent_$1.stat) \
-$(foreach module,$(ZEPHYR_$(AGENT_UPPER_CASE)_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_$(AGENT_UPPER_CASE)_DIR), $(ZEPHYR_$(AGENT_UPPER_CASE)_NAME)))
-endef
 
 ifdef BR2_PACKAGE_ZEPHYR_ARM_AGENT_HBM
-ZEPHYR_HBM_MODULES = eeprom mbedtls boot_info i2c-indirect soc_init
-$(call CREATE_ZEPHYR_MARGIN_TOOL_AGENT,hbm, $(HBM_EXTRA_CONF))
+ZEPHYR_DEPENDENCIES += zephyr-agent-hbm
 endif
 
 ifdef BR2_PACKAGE_ZEPHYR_ARM_AGENT_SERDES
-ZEPHYR_SERDES_MODULES = boot_info fdt i2c-indirect mbedtls soc_init
-$(call CREATE_ZEPHYR_MARGIN_TOOL_AGENT,serdes, $(PCIE_EXTRA_CONF))
+ZEPHYR_DEPENDENCIES += zephyr-agent-serdes
 endif
 
 ifdef BR2_PACKAGE_ZEPHYR_ARM_AGENT_PCIE
-ZEPHYR_PCIE_MODULES = boot_info i2c-indirect mbedtls soc_init
-$(call CREATE_ZEPHYR_MARGIN_TOOL_AGENT,pcie, $(SERDES_EXTRA_CONF))
+ZEPHYR_DEPENDENCIES += zephyr-agent-pcie
 endif
 
-ifeq ($(ZEPHYR_UNIT_TEST),1)
-ZEPHYR_UTEST_DIR = build-utest
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_UTEST_DIR)
-ZEPHYR_CONF_OPTS_UTEST += -DBOARD=hl_$(DEVICE_NAME)_mgmt_$(HL_TARGET)
-ZEPHYR_CONF_OPTS_UTEST += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_CONF_OPTS_UTEST += -DZEPHYR_MODULES=$(EMBEDDED_DIR)/zephyr-modules/eeprom;$(EMBEDDED_DIR)/zephyr-modules/boot_info
-ifneq ($(ZEPHYR_UNIT_TEST_DIR),)
-ZEPHYR_CONF_OPTS_UTEST += $(EMBEDDED_DIR)/zephyr/$(ZEPHYR_UNIT_TEST_DIR)
-else
-ZEPHYR_CONF_OPTS_UTEST += $(EMBEDDED_DIR)/zephyr/samples/hello_world
-endif
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,test,$(ZEPHYR_UTEST_DIR)) $(ZEPHYR_CONF_OPTS_UTEST) -B $(ZEPHYR_UTEST_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_UTEST_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_UTEST_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_utest.elf
-endif
-
-# Arguments:
-# - module (mbedtls, eeprom, etc.)
-define ZEPHYR_COMPONENT_MODULE_VERSION_HOOK
-	$(eval MODULE_NAME_UPPER_CASE=$(shell echo $1 | tr a-z A-Z))
-	$(eval MODULE_NAME_UPPER_CASE=$(strip $(MODULE_NAME_UPPER_CASE)))
-
-	$(call HL_VERSION_GEN, ${${COMPONENT_NAME_UPPER_CASE}_${MODULE_NAME_UPPER_CASE}_SRCDIR}, ${${COMPONENT_NAME_UPPER_CASE}_${MODULE_NAME_UPPER_CASE}_DIR}, ${${COMPONENT_NAME_UPPER_CASE}_${MODULE_NAME_UPPER_CASE}_NAME})
+define ZEPHYR_CONFIGURE_CMDS
+	$(call MESSAGE, "[TARGET] = $(HL_DEV) ZEPHYR_CONFIGURE_CMDS")
 endef
 
-# Arguments:
-# - component (zephyr apps and plugins)
-define ZEPHYR_COMPONENT_VERSION_HOOK
-	$(eval COMPONENT_UPPER_CASE=$(shell echo $1 | tr a-z A-Z))
-	$(eval COMPONENT_UPPER_CASE=$(strip $(COMPONENT_UPPER_CASE)))
-	$(eval COMPONENT_NAME=${ZEPHYR_${COMPONENT_UPPER_CASE}_NAME})
-	$(eval COMPONENT_NAME_UPPER_CASE=$(shell echo ${COMPONENT_NAME} | tr a-z A-Z))
-	$(eval COMPONENT_NAME_UPPER_CASE=$(strip $(COMPONENT_NAME_UPPER_CASE)))
-
-	$(call HL_VERSION_GEN, ${ZEPHYR_${COMPONENT_UPPER_CASE}_SRCDIR}, ${ZEPHYR_${COMPONENT_UPPER_CASE}_DIR}, ${COMPONENT_NAME})
-
-	$(foreach module,$(ZEPHYR_$(COMPONENT_UPPER_CASE)_MODULES),$(call ZEPHYR_COMPONENT_MODULE_VERSION_HOOK,$(module)))
+define ZEPHYR_BUILD_CMDS
+	$(call MESSAGE, "[TARGET] = $(HL_DEV) ZEPHYR_BUILD_CMDS")
 endef
 
-define ZEPHYR_SUBPACKAGES_VERSION_GEN
-	$(foreach component,$(ZEPHYR_COMPONENTS_LIST),$(call ZEPHYR_COMPONENT_VERSION_HOOK,${component}))
+define ZEPHYR_INSTALL_TARGET_CMDS
+	$(call MESSAGE, "[TARGET] = $(HL_DEV) ZEPHYR_DEPENDENCIES $(ZEPHYR_DEPENDENCIES)")
 endef
 
-zephyr-configure: $(ZEPHYR_DEPENDENCIES)
-zephyr-menuconfig zephyr-guiconfig: | zephyr-configure
-	$(TARGET_MAKE_ENV) $(MAKE) -C $(ZEPHYR_BUILDDIR)/$(ZEPHYR_CONF_DEFAULT) $(patsubst zephyr-%,%,$@)
-	rm -f $(ZEPHYR_BUILDDIR)/.stamp_{configured,built}
-
-$(eval $(cmake-package))
+$(eval $(generic-package))
